Fix security-enabled delete handling for nonexistent resources.

bowrocker opened the pull request:
Addresses issues:

https://github.com/basho/riak_kv/issues/918
https://github.com/basho/riak_kv/issues/917

bowrocker committed:
Fix security-enabled delete handling for nonexistent resources.

macintux commented on src/riak_kv_wm_object.erl:
```diff
@@ -264,14 +264,15 @@ forbidden(RD, Ctx=#ctx{security=Security}) ->
                 {true, _} ->
-                    case Perm of
-                        "riak_kv.get" ->
+                    case (Perm == "riak_kv.get" orelse
```
Almost entirely inconsequential, but I've seen commentary that `=:=` should always be used except when `==` is desired to allow for integer/float comparisons

kellymclaughlin replied:
I think either is ok here since we are definitely comparing two strings and the result will be the same.

kellymclaughlin commented:
I verified that this fix resolves the reported issues using the changes in [riak_test 612](https://github.com/basho/riak_test/pull/612). 

I suggested to Jon to also do a bit of code cleanup related to some of these changes so I will withhold  final approval until that is completed.

bowrocker committed:
Refactor forbidden/2 and resource validation helper functions.

kellymclaughlin commented on src/riak_kv_wm_object.erl:
```diff
@@ -219,68 +219,61 @@ is_authorized(ReqData, Ctx) ->
+    validate_bucket_type(RD, Ctx).
+
+-spec method_to_perm(atom()) -> string().
+method_to_perm(Method) when Method == 'POST' ->
```
I would rewrite these function clauses to pattern match on the different method atoms in the parameter instead of using the guard to check. 

_e.g._ `method_to_perm('POST') ->`

kellymclaughlin replied:
How about moving this function to `riak_kv_wm_utils` and using it here and in `riak_kv_wm_props`. There may be other places that case statement makes an appearance as well.

bowrocker committed:
Further refactorings to forbidden/2.

- Move method_to_term to riak_kv_wm_utils.
- Put resulting and unexplained dialyzer error into ignore file.
- Rename validate_unforbidden.

bowrocker committed:
Fix R15 dialyzer issue in riak_kv_wm_utils.

- Remove the case clause after the http_uri:parse/1, it has been
deprecated.
- Remove the entry for the dialyzer warning in the ignore file.

bowrocker committed:
Fix whitespace.

kellymclaughlin commented:
Refactoring looks good and dialyzer output is clean.

:+1:

bowrocker commented:
@borshop merge

kellymclaughlin commented:
+1 40b04a1
